### PR TITLE
Docs: routing DSL compared with Play routes

### PR DESCRIPTION
**Pull request by ennru**
Add a page in the routing docs which shows the examples from the Play docs in Akka HTTP notation with accompanying tests.

References #2980

### CODE REVIEW

File: docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      val itemParameterList: Route =
+        // #parameter-list
+        // TODO:
+        // Note that implicit conversions are not applicable because they are ambiguous:
+        //  both method augmentString in object Predef of type (x: String)scala.collection.immutable.StringOps
+        //  and method _string2NR in trait ToNameReceptacleEnhancements of type (string: String)akka.http.scaladsl.common.NameReceptacle[String]
+        //  are possible conversion functions from String("item") to ?{def *: ?}
+        //         path("api" / "list-items")(parameters("item".*) { items =>
+        get {
+          path("api" / "list-items")(parameters(ToNameReceptacleEnhancements._string2NR("item").*) { items =>
```
Comment (raboof) [inline docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala:124]: If we want to keep PR's independent, we could keep showing the symbol syntax here for now until we introduce syntax for this that works with strings?

File: docs/src/main/paradox/routing-dsl/index.md [hunk 2]
```diff
@@ -69,6 +70,10 @@ the Routing DSL will look like:
 
 @@@
 
+## Compared with Play framework routes
+
+If you have been using Play framework's routes file notation before this @ref[Play! comparison](play-comparison.md) may help you to get started with Akka HTTP routing.
```
Comment (ignasi35) [inline docs/src/main/paradox/routing-dsl/index.md:75]:
```suggestion
If you have been using Play framework's routes file notation before this @ref[Play comparison](play-comparison.md) may help you to get started with Akka HTTP routing.
```

I don't think we use the `!` anymore.

File: docs/src/main/paradox/routing-dsl/play-comparison.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+If you have been using @scala[[Play's routes file syntax](https://www.playframework.com/documentation/2.8.x/ScalaRouting#The-routes-file-syntax)]@java[[Play's routes file syntax](https://www.playframework.com/documentation/2.8.x/JavaRouting#The-routes-file-syntax)] earlier, this page may help you to use the Akka HTTP routing DSL.
+
+Where Play routes require to specify the HTTP method for every entry, in Akka HTTP it may be specified on different levels of the `Route`.
+
+Both Play and Akka HTTP choose the first matching route within the routes file/routes definition. In Play routes are listed with one route per line, in Akka HTTP multiple routes must be concatenated with the `concat` method.
+
+### Static path
+
+For example, to exactly match incoming `GET /clients/all` requests, you can define this route in Play!.
```
Comment (ignasi35) [inline docs/src/main/paradox/routing-dsl/play-comparison.md:11]:
```suggestion
For example, to exactly match incoming `GET /clients/all` requests, you can define this route in Play.
```

etc...

File: docs/src/test/java/docs/http/javadsl/server/testkit/PlayRoutesComparisonTest.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    static String apiItems(List<String> items) {
+        return items.stream().collect(Collectors.joining(","));
+    }
+
+    static class Routes extends AllDirectives {
+
+        Route clientsAll() {
+            return
+                    // #fixed
+                    get(() -> path(segment("clients").slash("all"), () -> complete(Clients.list())));
```
Comment (ignasi35) [inline docs/src/test/java/docs/http/javadsl/server/testkit/PlayRoutesComparisonTest.java:55]: Screenshot 2020-04-17 at 16 43 59


I found all these one-liners (in both tests and main code for Java) hard to read. I think that, even if it's not java idiomatic and feels for scala-esque, somthing like the following is more readable:

```suggestion
  get(() -> 
    path(segment("clients").slash("all"), 
      () -> complete(Clients.list())
    )
  );
```

I mean, if we're writing these docs for someone coming from Play, I think it's visually easier to parse this format then the one-liners and gently introduces the reader to a slightly different format.

File: docs/src/main/paradox/routing-dsl/routes.md [hunk 5]
```diff
@@ -88,8 +88,8 @@ of either the incoming request, the outgoing response or both
  * Route filtering, which only lets requests satisfying a given filter condition pass and rejects all others
  * Route chaining, which tries a second route if a given first one was rejected
 
-The last point is achieved with the concatenation operator `~`, which is an extension method that becomes available
-when you `import akka.http.scaladsl.server.Directives._`.
+@scala[The last point is achieved with the concatenation operator `~` (or the `concat` method), which is an extension method that becomes available
```
[the remark this answers is not in the document]
Comment (ennru) [reply docs/src/main/paradox/routing-dsl/routes.md]: I removed the `~` altogether from here.

File: docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala (hunk 1 above)
Comment (ennru) [reply docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala]: Did that now. Other PR incoming...

File: docs/src/test/java/docs/http/javadsl/server/testkit/PlayRoutesComparisonTest.java (hunk 4 above)
Comment (ennru) [reply docs/src/test/java/docs/http/javadsl/server/testkit/PlayRoutesComparisonTest.java]: Reformatted.

File: docs/src/main/paradox/routing-dsl/directives/parameter-directives/parameters.md [hunk 6]
```diff
@@ -115,6 +115,9 @@ Java
 
 ### Parameter with required value
 
+@scala[The `!` decorator makes the route match only if the parameter contains the specified value.]
```
Comment (jrudolph) [inline docs/src/main/paradox/routing-dsl/directives/parameter-directives/parameters.md:118]: Does that need to be updated after #3082?

File: docs/src/main/paradox/routing-dsl/play-comparison.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
+# Compared with Play routes
+
+If you have been using @scala[[Play's routes file syntax](https://www.playframework.com/documentation/2.8.x/ScalaRouting#The-routes-file-syntax)]@java[[Play's routes file syntax](https://www.playframework.com/documentation/2.8.x/JavaRouting#The-routes-file-syntax)] earlier, this page may help you to use the Akka HTTP routing DSL.
+
+Where Play routes require to specify the HTTP method for every entry, in Akka HTTP it may be specified on different levels of the `Route`.
```
Comment (jrudolph) [inline docs/src/main/paradox/routing-dsl/play-comparison.md:5]: Not quite sure what that means and why it is the first thing to say here :)

**Comment by ennru** [comment]
This will require some updated if we agree on #3082

File: docs/src/main/paradox/routing-dsl/routes.md [hunk 8]
```diff
@@ -173,7 +173,7 @@ A sealed route has these properties:
 As described in @ref[Rejections](rejections.md) and @ref[Exception Handling](exception-handling.md),
 there are generally two ways to handle rejections and exceptions.
 
- * Bring rejection/exception handlers @scala[`into implicit scope at the top-level`]@java[`seal()` method of the @scala[@scaladoc[Route](akka.http.scaladsl.server.index#Route=akka.http.scaladsl.server.RequestContext=%3Escala.concurrent.Future[akka.http.scaladsl.server.RouteResult])]@java[@apidoc[Route]]]
+ * Bring rejection/exception handlers @scala[into `implicit` scope at the top-level]@java[`seal()` method of the @apidoc[Route]]
```
Comment (jrudolph) [inline docs/src/main/paradox/routing-dsl/routes.md:176]: That's not correct even for Scala any more.

File: docs/src/main/paradox/routing-dsl/play-comparison.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
+# Compared with Play routes
+
+If you have been using @scala[[Play's routes file syntax](https://www.playframework.com/documentation/2.8.x/ScalaRouting#The-routes-file-syntax)]@java[[Play's routes file syntax](https://www.playframework.com/documentation/2.8.x/JavaRouting#The-routes-file-syntax)] earlier, this page may help you to use the Akka HTTP routing DSL.
+
```
Comment (jrudolph) [inline docs/src/main/paradox/routing-dsl/play-comparison.md:4]: I think we need an overview section about the more general differences between play and akka-http routes.

 * internal vs. external DSL
 * what is covered (and what not) by play route files vs. what is covered by Akka HTTP routing

File: docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+    def get(id: Long): String = "clientB"
+  }
+
+  "Play examples" should {
+
+    "show a fixed URL" in {
+      val clientsAll: Route =
+        // #fixed
+        (get & path("clients" / "all")) {
```
Comment (jrudolph) [inline docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala:28]: While you could write it that way, I think we should recommend a different style in our new best practices guide. There, we would prefer matching on path first and then on method. But in general, those might not be in the same line of code. We could still show them in the same line of code but would then point back to that guide about how you would do it more generally.

File: docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+      Get("/clients/all") ~> clientsAll ~> check {
+        responseAs[String] shouldEqual "clientA,clientB,clientC"
+      }
+      // #fixed-test
+    }
+
+    "show reading a long" in {
+      val clientById: Route =
+        // #long
+        (get & pathPrefix("client" / LongNumber)) { id =>
```
Comment (jrudolph) [inline docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala:43]: Probably `path` here but the same comment as above applies.

File: docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+      }
+      // #long-test
+    }
+
+    "use multiple segments" in {
+      def download(name: String) = s"$name: file contents"
+
+      val files: Route =
+        // #segments
+        (get & pathPrefix("files" / Segments)) { names =>
```
Comment (jrudolph) [inline docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala:60]: You would probably use `Remaining` or `RemainingPath` instead here.

File: docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+      }
+      // #segments-test
+    }
+
+    "require a parameter" in {
+      def getPage(name: String) = s"The requested [$name]."
+
+      val pageParameter: Route =
+        // #mandatory-parameter
+        (get & parameter("page")) { page =>
```
Comment (jrudolph) [inline docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala:78]: We should add a path here as well to make it consistent with Play.

File: docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+      // #mandatory-parameter-test
+    }
+
+    "accept a parameter" in {
+      def listAll(version: Option[String]) = version.fold("ff")(_ => "aa,bb,cc")
+
+      val optionalPageParameter: Route =
+        // #optional-parameter
+        get {
+          path("api" / "list-all")(parameter("version".?) { version =>
```
Comment (jrudolph) [inline docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala:96]: Let's either use proper nesting with one directive per line, or use `&`. I.e.

```suggestion
          (path("api" / "list-all") & parameter("version".?)) { version =>
```

File: docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala (hunk 10 above)
Comment (ignasi35) [reply docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala:28]: Hmm, that's a good idea. If, in the very fisrt example we introduced:

```
// Play routes file
GET				/clients/all    controllers.getAll()

// Akka-HTTP
(get & path("clients" / "all")) {
  ...
}

// idiomatic Akka-HTTP
path("clients" / "all"){
  get {
    ...
  }
}
```

So users can see it's possible to do a more direct conversion but we specify, early on in the page, that the idiomatic Akka-HTTP approach is to match on the path before matching on the method.

Comment (ennru) [reply docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala:26]: I'd like to keep the focus on the direct comparison and link to the style guide.

Comment (ignasi35) [reply docs/src/test/scala/docs/http/scaladsl/server/PlayRoutesComparisonSpec.scala:26]: Ah, good idea! Introducing the style guide early is a better approach.

File: docs/src/main/paradox/routing-dsl/play-comparison.md (hunk 9 above)
Comment (ennru) [reply docs/src/main/paradox/routing-dsl/play-comparison.md:4]: I added a more general introduction, but I don't want to go too deep as I think the main purpose of this is the side-by-side illustration.

**Comment by jrudolph** [comment]
Woops, that should have been squashed, @ignasi35. ;)

**Pull request merged**